### ISSUE TITLE
Feature: Support DoT (DNS over TLS) RFC 7858

### ISSUE THREAD

**Issue by bradh352** [c-ares/c-ares#818]
https://datatracker.ietf.org/doc/html/rfc7858

DNS over TLS provides some layer of privacy by using an encrypted channel.  Arguably DoH (DNS over HTTPs) (#612) provides more privacy as it is indistinguishable from other HTTPS traffic, however we need to basically implement DoT first before we can get to DoH, so we might as well split it out as its own feature.

Also, we should implement TCP FastOpen (#815) before this to reduce the latency involved in connection establishment, in combination with 0-RTT TLSv1.3.

**Comment by createyourpersonalaccount** [comment]
I would like to try this one. I'm thinking I'll start with OpenSSL 3.0 and go from there, see if I can get it done. Just wondering if anyone is already working on this?

**Comment by bradh352** [comment]
I'm almost done with this already.  I just have to hook in a couple more things then enhance the test system for testing it properly.  I really should just finish it up, I haven't really touched the branch I'm working on in a few months as I got caught up in exploring [SONiC](https://github.com/sonic-net/sonic-buildimage) and submitting patches to fix things. :)

**Comment by createyourpersonalaccount** [comment]
Nice! It was a good idea to check before starting this...

**Comment by bradh352** [comment]
@createyourpersonalaccount 

I guess I should clarify a bit more on this.  My branch is here: https://github.com/bradh352/c-ares/tree/DoT

The part I'm almost done with is what I'd consider Phase 1, which only supports manually specified DoT servers using the URI schemes I introduced in preparation in PR #882.  

Phase 2 would entail splitting priority groups for servers, as if DoT (or in the future DoH) servers are specified, it shouldn't route queries to non-secure servers unless the non-secure servers aren't fully *configured*.  Some systems will only provide a DNS IP and not a DNS hostname (or vice-versa), and we may need to look up the opposite via non-secure servers during system initialization (or during first real query) in order to do things like certificate validation.

Phase 3 would be reading the configuration from the various OS's.  Windows looks like the most straight forward.  Linux we'd likely have to bypass systemd-resolved and read its config files rather than /etc/resolv.conf (which I wouldn't mind doing anyhow).  Haven't seen any other OS's introduce support with their system configs at all.

Phase 4 would be optimization / configuration flexibility.  Meaning allowing users to specify to not validate the hostname, or maybe specifying a custom CA certificate to use to validate the server, or maybe even client certificates for authenticating with the upstream server.

So if you're looking on working towards this feature, you could look at tackling some of the other phases.  Personally I see parts of Phase 2 overlapping with #642 (domain-specific servers) in grouping servers for specific cases.  Actually, it also overlaps a bit with Phase 3 at least on Linux as we'd need to read from systemd-resolved's config directly for domain-specific servers since /etc/resolv.conf doesn't have that concept (this too would utilize PR #882 's URI schemes).

**Comment by createyourpersonalaccount** [comment]
@bradh352 thanks for the additional info. I'll look into this Phrase 2/3 then after I get up to speed with your branch.

**Comment by bradh352** [comment]
@createyourpersonalaccount you'll just need to look at the diff from main from my branch, the commit history on it is pretty bad because I derived multiple other PRs from that branch (#876, #875, #857) in order to not make one massive change on the way to DoT.

**Comment by marjakm** [comment]
I'm interested in this feature - phase 1 would be good enough for me. I'm considering trying to bring the https://github.com/bradh352/c-ares/tree/DoT branch up to date with upstream main, but I haven't actually used c-ares and know nothing about the codebase. Would this be a relatively straightforward task or not?

Has there been any progress since the this branch?

Context: I like the idea of having a publicly accessible DoT-only DNS server that does client cert authentication (mtls) and only servers DNS records to machines that I control.
